src/evict/evict_lru.c

```diff
@@ -773,14 +773,14 @@ __evict_clear_walk(WT_SESSION_IMPL *session, bool count_stat)
 	if ((ref = btree->evict_ref) == NULL)
 		return (0);
 
-	if (count_stat)
-		WT_STAT_CONN_INCR(session, cache_eviction_walks_abandoned);
+	WT_STAT_CONN_INCR(session, cache_eviction_walks_abandoned);
 
 	/*
-	 * Clear evict_ref first, in case releasing it forces eviction (we
-	 * assert we never try to evict the current eviction walk point).
+	 * Clear evict_ref before releasing it in case that forces eviction (we
+	 * assert that we never try to evict the current eviction walk point).
 	 */
-	btree->evict_ref = NULL;
+	WT_PUBLISH(btree->evict_ref, NULL);
```

Comment:
@agorrod, I think this change is wrong.

`WT_PUBLISH` is a write-barrier before the assignment, and what we need here is to ensure the update of `btree->evict_ref` be done before any subsequent reads of the value.  I think this should be:
```
btree->evict_ref = NULL;
WT_FULL_BARRIER();
```
I've made this same mistake. @michaelcahill, how about:
```
/*
 * Write a shared location and guarantee that subsequent reads do not see any
 * earlier state.
 */
#define WT_ORDERED_WRITE(v, val) do {
        (v) = (val);
        WT_FULL_BARRIER();
} while (0)
```